Fire Rework

west3436 opened the pull request:
<!--
Pull requests must be atomic. Change one set of related things at a time.
Test your changes. PRs that were not tested will not be accepted.

You can self-label your PR. See https://ss13.moe/wiki/index.php/Guide_to_Writing_a_Pull_Request -->

## What this does
<!-- Describe here all changes included in the PR. -->
<!-- If the PR addresses existing issues, here is where you would write "Closes #99999". See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->
Reworks the way objects burn while maintaining the existing gas combustion mechanics.

### List of changes:
**Gameplay:**
- Temperature changes from fire now follow semi-realistic calculations, resulting in fires that grow at an exponential rate, starting very slowly and growing into a conflagration once material autoignition temperatures are met.
- Flame temperatures are limited to the maximum flame temperature of the burning material. As a result, burning objects can no longer heat rooms to be as hot as the surface of the sun. Plasma and oxygen burning does not have a temperature cap.
- The fire triangle is now modelled directly: removing the fuel, oxygen, or heat from a fire scene will slow down or completely put out the fire.
- Items approaching their autoignition temperature will begin to emit smoke particles.
- Things that have been exposed to fire or burnt will be covered in ash. Ash can be cleaned with space cleaner.
- Beware of backdrafts - care must be taken when entering a room full of smoke without any visible fire. Reintroducing oxygen into the room could cause spontaneous combustion!
- Smoke now scales inversely with heat - hotter fires will produce less smoke.
- Autoignition has been restored for flammable objects.
- Autoignition has been implemented for flammable turfs.

**Refactor:**
- **Added w_class and w_type values for numerous objects.**
- Flammability properties for all items are now drawn from an underlying material type defined by their w_class and w_type.
- Contributors now only have to set "flammable = TRUE" for any items they want to burn instead of providing an autoignition temperature and an, often subjective, fire_fuel amount.
- Heat generation, fire oxygen consumption, and fire spread rate can all be tweaked easily in code/ZAS/NewSettings.dm.

**Performance:**
- Fire color + light will now refresh every 8-12 seconds rather than every second to improve performance.

## Why it's good
<!-- Explain why you think these changes are good. -->
The bulk of fire code was created 10-12 years ago. While I have previously contributed some small changes to the way fire affects things, the underlying system has not changed in any significant way until #35306 was merged and _fixed_ autoignition of objects. As things now were actually catching on fire, many of the flaws of the ancient fire code began to come to light; creating hilarities such as supernova donut piles, cigarette smokers being atmos techs' public enemy number one, wizard fireballs setting the entire station on fire, rods creating a line of fires wherever they struck (okay this one was pretty cool), and infinite chains of fires if a single room got hot enough to heat its neighbors. This fun rightfully came to an end in #36028 as continuously slapping bandages on the old code simply was not enough to make it a fun system to play with.

This rework aims to restore the fun opportunities created by autoignition by fixing the underlying system entirely. Fires will now grow at a slower rate than the baseline but can still grow into a conflagration if left unchecked for too long. This system aims to model the fire triangle such that a dedicated firefighter or firefighting team can put out a fire by either removing the fuel from the room, cooling the room down, or removing the oxygen from the room. Fires will no longer be round-ending catastrophes spreading from room to room unchecked, but will still be a significant danger to crewmembers in the area if an effort isn't taken to at least cordon off the affected region until the area has cooled off.

Closes #36128.
Closes #35874.
Closes #35928.

## Visuals
![smoke](https://github.com/vgstation-coders/vgstation13/assets/66280799/10ce140d-5a1a-468e-97c0-08e69d972d07)
Items beginning to give off smoke as they approach their autoignition temperature.

[Fire Burn.webm](https://github.com/vgstation-coders/vgstation13/assets/66280799/a76f8abb-23a2-4d62-88ed-7fd8d2d6ad94)
Video of the bar burning down after a chair was ignited.

![result](https://github.com/vgstation-coders/vgstation13/assets/66280799/f76a75d7-ac86-48d6-9810-afa4fba06f56)
Atmos in the bar following the pictured burn (no longer the temperature of the sun).

!![image](https://github.com/vgstation-coders/vgstation13/assets/66280799/ddc17a79-0298-4489-887c-3e29740773e8)
The bar covered in ash after a rapid flashover put itself out due to lack of oxygen.

## Future
Planning for future additions to this system if it is merged including:
- Specific fire extinguishers for specific fuel types
- Random fire event
- Improving the ash visuals including ash footsteps
- Advanced Fire Alarms which trigger siphon mode/stop adding oxygen to the fire
- Manually igniting flammable objects using a hot tool
- Cryotheum and fire interactions (inverse fire??)
- Worn or held items or clothing catching on fire

## Changelog
<!-- See https://ss13.moe/wiki/index.php/Guide_to_Writing_a_Pull_Request -->
:cl:
 * experiment: Reworked fire.
 * bugfix: Fire color and light will no longer lag the game by refreshing every second.
 * tweak: Fires caused by burning objects will no longer reach unrealistic flame temperatures.

Eneocho commented on __DEFINES/setup.dm:
```diff
@@ -1129,17 +1129,49 @@ var/default_colour_matrix = list(1,0,0,0,\
+#define SPECIFIC_HEAT_WOOD 18.0
+#define SPECIFIC_HEAT_PLASTIC 46.4 // Polypropylene
+#define SPECIFIC_HEAT_FABRIC 19.4 // Cotton (https://www.researchgate.net/publication/363517205_Analysis_of_the_Flammability_and_the_Mechanical_and_Electrostatic_Discharge_Properties_of_Selected_Personal_Protective_Equipment_Used_in_Oxygen-Enriched_Atmosphere_in_a_State_of_Epidemic_Emergency)
+#define SPECIFIC_HEAT_WAX 42.0 // Paraffin wax
```
shouldn't it be beeswax?
https://www.engineeringtoolbox.com/specific-heat-solids-d_154.html

west3436 replied:
Let's not split hairs over this - I've had a much easier time finding material properties for paraffin than I have for beeswax. Space bees are just built different and produce paraffin 🥴

Eneocho replied:
I literally googled "material properties of beeswax" and found this:
https://sci-hub.se/https://www.tandfonline.com/doi/pdf/10.1179/026708302225001714 (mechanical properties of beeswax), the first link I posted has the specific heat, wikipedia has an approximate chemical formula https://en.wikipedia.org/wiki/Beeswax#cite_ref-10 (C15H31COOC30H61) (with reference). It apparently has no autoignition, instead just melting then flashing.
https://fscimage.fishersci.com/msds/02556.htm a material data sheet for beeswax.

Eneocho commented on __DEFINES/setup.dm:
```diff
@@ -1129,17 +1129,49 @@ var/default_colour_matrix = list(1,0,0,0,\
+#define MOLECULAR_WEIGHT_WOOD 0.16 // simplified to C6H10O5 (cellulose)
+#define MOLECULAR_WEIGHT_PLASTIC 0.042 // C3H6 ((poly)propylene)
+#define MOLECULAR_WEIGHT_FABRIC 0.16 // simplified to C6H10O5 (cellulose)
+#define MOLECULAR_WEIGHT_WAX 0.35 // C25H52 (paraffin wax)
```
ditto

west3436 replied:
ditto

boy2mantwicethefam commented:
please.........................................no more thorough reworks that have a lot of potential breaking points.............................

Eneocho commented:
Hang on, lets see where this goes. Fire right now is extremely overtuned, a full rework might be precisely what we need.

CrazyAmphibian commented:
so, what's the goal of the rework?

west3436 commented:
> please.........................................no more thorough reworks that have a lot of potential breaking points.............................

Slapping random tweaks onto the existing system isn't going to work; the entire system needs to be redone.


> so, what's the goal of the rework?

Primary goal: 
- Make burning items more realistic in the sense that a pile of donuts (the classic example) doesn't make the station go supernova. I am not going for full realism and I am not doing it just for the sake of realism - I just want fire to "feel" right and not wildly overtuned as it does now.

Bonus goals:
- The new system brings some sanity to the way fires burn; a burning object will start a full-tile fire, the fire will then consume "thermal mass" from the burning object, the tile itself (if flammable), and any reagents on the tile or in containers on the tile and turn it into heat. This system attempts to conserve energy in a realistic manner which will prevent infinitely-heating fires if a constant fuel source is provided.
- The new system will include flammable reagents burning, so a beaker full of welding fuel or eventually the deep fryer can cause a big fire.
- The new system will be much easier to use for coders adding things; provided a coder properly defines the material type, then they just have to define the var "flammable" as TRUE or FALSE on the item instead of having to define its autoignition temperature and/or a subjective fire_fuel value.
- The new system can easily be tweaked to eventually have other materials like metal burn if the proper conditions are met (pure oxygen, high pressure).

hacker-on-steroids commented on __DEFINES/setup.dm:
```diff
@@ -1129,16 +1129,50 @@ var/default_colour_matrix = list(1,0,0,0,\
+// FIRE
+//////////////////
+
+// specific heats for burning objects, assuming perfect combustion (MJ/kg)
```
is this supposed to be specific heat capacity, in J/kgK?
or fuel burn value?

if B please name it `fuel_value` or literally anything else because it's extremely similar to the chemistry specheatcaps, which have been notoriously retarded until the frenchman fixed them recently

west3436 replied:
It's actually the heat of combustion which is measured in MJ/kg. I'll rename it accordingly so we don't get it mixed up with specific heat capacity.

hacker-on-steroids commented on code/ZAS/Fire.dm:
```diff
@@ -43,101 +83,151 @@ Attach to transfer valve and open. BOOM.
+		extinguish()
+		return burn_products
+
+	//smoke density increases with burnrate and temperature
```
not a fireman but i'd reverse the temperature
higher temperature usually means more complete combustion which means less smoke

west3436 replied:
Good call

hacker-on-steroids commented on code/ZAS/Fire.dm:
```diff
@@ -43,101 +83,151 @@ Attach to transfer valve and open. BOOM.
-		smoke.start()
+	var/heat_out = 0 //MJ
+	var/oxy_used = 0 //mols
+	var/list/burn_products = list(heat_out,oxy_used)
```
this proc will always return [0,0]
you'd need to make them objects for it to use references instead of copying

hacker-on-steroids commented on code/ZAS/Fire.dm:
```diff
@@ -43,101 +83,151 @@ Attach to transfer valve and open. BOOM.
+	//rate at which energy is consumed from the atom and delivered to the fire
+	//burnrate = 1 at 20C with standard oxy concentration
+	//provides the "heat" and "oxygen" portions of the fire triangle
+	var/burnrate = oxy_ratio >= MINOXY2BURN ? (oxy_ratio/(MINOXY2BURN + rand(-0.02,0.02))) * (temperature/T20C) : 0
```
i would add a pressure component and reduce the effect on the temperature one
and possibly make the multipliers config settings or better yet adjustable ingame, then fiddle with them to get good results

actually now that i think about it molar density * temperature is just pressure/8.31 isn't it?

west3436 replied:
The oxy ratio drops pretty rapidly which is why I give temperature a decent impact. This whole equation is made up bullshit but as mentioned in the comments, it allows people to use the fire triangle to reduce the burning rate of the fire or put it out completely.

hacker-on-steroids replied:
well current solid fires neither use oxygen nor make co2
i'm assuming with this rework you'll do the atmos adjustments after this proc finishes and you get the burn_products returned because i don't see any gas adjustment with solids so far

west3436 replied:
Correct - atmos changes will be done in zburn or fire_act but I haven't got there yet. I'll work on it more this upcoming week.

hacker-on-steroids replied:
i believe this now simplifies to roughly `some constant` * `oxygen partial pressure` ~~* `pressure`~~
gonna be interesting to see how this works ingame
long as it doesn't turn cigarette butts and donuts into weapons of mass destruction it's fine by me

west3436 replied:
Yeah pressure shouldn't be directly factored in here; fires can still burn in low pressure. oxy_ratio * temperature != pressure/8.31 as the oxy_ratio is the molar fraction of oxygen only

west3436 commented:
Ready for code review - will capture and post some video/image comparisons between this, current fire, and pre-autoignition-nerf fire this evening.

CrazyAmphibian commented:
backdrafts sound like a cool (and realistic) challenge to deal with, even if a few inflatable barriers are the solution.

smashedandslammed commented:
Revert all hinacode before doing this shit

west3436 commented:
> Revert all hinacode before doing this shit

If you read the OP, the only hinacode was the fix to autoignition via the addition of burnable.dm. I've modified burnable.dm as well as the heat produced by fires, so many of the pains of autoignition felt prior to the recent nerf will not be felt this time around (ie a stack of papers won't make the entire station 6000K).

Kurfursten commented:
How is autoignition now being received for an item if it's just flammable as a var? For example, what tells teeth when to ignite?

west3436 commented:
> How is autoignition now being received for an item if it's just flammable as a var? For example, what tells teeth when to ignite?

Autoignition occurs in burnable.dm which is checked I believe every 7 seconds.

Kurfursten reviewed (commented):
I have looked over 179 files and the changes seem fine

I'd like Exxion to look at Fire.dm and if he likes it then this is clear to merge

hacker-on-steroids commented on code/ZAS/Fire.dm:
```diff
@@ -1,34 +1,118 @@
+				thermal_material = new/datum/thermal_material/wax()
+			if(RECYK_BIOLOGICAL)
+				thermal_material = new/datum/thermal_material/biological()
+		if(!thermal_material)
```
logspam

`## WARNING: The department debit card was defined as flammable but was missing a 'w_type' definition; the department debit card marked as inflammable for this round`

hacker-on-steroids replied:
also, if my understanding of english is correct, "inflammable" means easily flammable which the department debit cards aren't

Eneocho replied:
https://www.youtube.com/watch?v=Q8mD2hsxrhQ

west3436 replied:
>logspam

Yeah going to axe that

>grammar

You are correct, "inflammable" is the wrong word

west3436 replied:
Fixed in latest commit

hacker-on-steroids commented on __DEFINES/setup.dm:
```diff
@@ -1132,16 +1132,53 @@ var/default_colour_matrix = list(1,0,0,0,\
 #define MT_REINIT 2
 
+//////////////////
+// FIRE
```
igniting a welding fuel puddle or anything else really seems to be impossible, neither dropping a lit welder, using an igniter or throwing an IED works
i haven't actually figured out a way to start a fire without the ignite() proc or popping a canned heat

worn clothes, ID cards and boxes inside backpacks will vanish in 1 tick if you pop a canned heat or walk into a hot room - you should definitely add a minimum burn time
PDAs aren't flammable but ID cards are, and as someone who doesn't combine the 2 i really hate this

boxes burning in your backpack drop contents on the floor instead of the backpack (they probably shouldn't burn while inside the backpack anyway)
you should make a visible_message when an item turns into ash

plasma gas doesn't ignite in the TEG any more
clothes burn even if you have a hardsuit on

hacker-on-steroids replied:
[13:37:22] Runtime in Fire.dm,202: Cannot execute null.return air().
  proc name: burnSolidFuel (/atom/proc/burnSolidFuel)
  src: the Bidi (/obj/item/clothing/mask/cigarette/bidi)
  src.loc: null

west3436 replied:
>igniting a welding fuel puddle or anything else really seems to be impossible, neither dropping a lit welder, using an igniter or throwing an IED works
i haven't actually figured out a way to start a fire without the ignite() proc or popping a canned heat

I've ran into this issue in previous iterations of this but thought I fixed it; I'll get that working again.

>worn clothes, ID cards and boxes inside backpacks will vanish in 1 tick if you pop a canned heat or walk into a hot room - you should definitely add a minimum burn time

Had a minimum burn time previously but it seemed inorganic; I agree that things disappearing in a single tick is even more inorganic so I will restore that.

>PDAs aren't flammable but ID cards are, and as someone who doesn't combine the 2 i really hate this

Might just make IDs fireproof as people got very upset about them being flammable previously.

>boxes burning in your backpack drop contents on the floor instead of the backpack (they probably shouldn't burn while inside the backpack anyway)

Feature - boxes recursively drop their contents when they are set on fire otherwise backpacks full of boxes of things will burn for ages.

>you should make a visible_message when an item turns into ash

Good idea, will implement

>plasma gas doesn't ignite in the TEG any more

Likely tied to the other issues with ignition

>clothes burn even if you have a hardsuit on

I have been debating on eliminating clothes burning while worn and I think I'll just strike that feature entirely for now until people get used to the new system.

>runtime

Thanks, will fix

Thanks for the quality review, got a lot of work to get this working still. Much appreciated

west3436 replied:
Had a bunch of incorrect comparisons for fire_protection and world.time which was preventing things from igniting; going to test this evening but the struggle with igniting things is likely fixed now

Exxion commented:
I'm working on an actual review of the file, but you should be aware that `firelightdummy` doesn't actually have to be added to `vis_contents` to work. It works just because it's in regular `contents`. It also causes hard dels all the fucking time due to being in `vis_contents` and not properly removed. I'd recommend just... removing all the code that adds it to `vis_contents`

Eneocho commented:
I'll give it a stress testing this weekend, I'm kinda busy lately.

west3436 commented:
@hacker-on-steroids not only did I have inverted comparisons for a fireproofing check but hotspot_expose() wasn't working like at all. It's all updated and functioning perfectly now. Thanks for the sanity checks

malfunctioning-ai commented:
So I see this PR mainly touches how fires are handled in regard to atoms, i.e. an attempt at re-doing what Hina did with autoignition and stuff and adjusting how the 
Some big notes:
- The burnableatoms list in the burnable subsystem NEVER currently removes any of these flammable atoms from it, even after that flammable atom is destroyed. This needs to happen even if it's as simple as `atom/Destroy -> if(flammable), burnableatoms -= this_atom`. In general there probably needs to be a more efficient data structure for burnableatoms (and for all the subsystems for this matter, but especially burnableatoms due to its size). Basic lists are array structures, which perform copies on removing an element and have O(n) search time. 
- I see you changed it so that each area has a specific limit on how much smoke can be spawned in it. Is this really the best method? I assume grabbing the area any turf is in is fast (or so I hope), but you should probably also scale the maximum smoke in an area to the size of an area at least.
- Have you considered slowing down non-gas fires like 10x? I saw your bar fire webm ended in like 30 seconds. I think it would be cool for it to slowly ramp up and burn for like 5 minutes at least to give an actual chance for "firefighting". Maybe don't apply that to gas fires for TEG purposes, or otherwise make it so plasmafires operate much faster than wood fires.
- I didn't actually check any of the hundreds of changes to flammability values on various items, I'm guessing you did them right.

west3436 commented on code/controllers/subsystem/burnable.dm:
```diff
@@ -42,14 +35,28 @@ var/list/atom/burnableatoms = list()
-#define MINOXY2BURN (1 / CELL_VOLUME)
 /atom/proc/checkburn()
-	if(on_fire)
+	if(flammable)
```
Good point; that can help with performance. Will implement with added documentation.

west3436 commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -162,6 +162,13 @@
+	return thermal_mass * amount
+
+/obj/item/stack/useThermalMass(var/used_mass)
+	thermal_mass -= used_mass / amount
```
Yeah the issue with the previous approach to this was that a stack of fire would burn for a very long time and heat the room up significantly. I'll think of a new approach and rework it.

west3436 commented:
Thanks for the review!

> * The burnableatoms list in the burnable subsystem NEVER currently removes any of these flammable atoms from it, even after that flammable atom is destroyed. This needs to happen even if it's as simple as `atom/Destroy -> if(flammable), burnableatoms -= this_atom`. In general there probably needs to be a more efficient data structure for burnableatoms (and for all the subsystems for this matter, but especially burnableatoms due to its size). Basic lists are array structures, which perform copies on removing an element and have O(n) search time.

Added the check in atom/Destroy for now to remove items. Talking with the elder coders to see how this can be optimized further.


> * I see you changed it so that each area has a specific limit on how much smoke can be spawned in it. Is this really the best method? I assume grabbing the area any turf is in is fast (or so I hope), but you should probably also scale the maximum smoke in an area to the size of an area at least.

In the best conditions there is only a 5% chance that a fire tick generates smoke, so I can put that probability check before area check to reduce the amount of times the area is grabbed. I'll play around with what looks best and make it scale with space. A better approach would be to put the smoke on its own controller and limit generation based on tick usage but that's a pretty hefty change.


> * Have you considered slowing down non-gas fires like 10x? I saw your bar fire webm ended in like 30 seconds. I think it would be cool for it to slowly ramp up and burn for like 5 minutes at least to give an actual chance for "firefighting". Maybe don't apply that to gas fires for TEG purposes, or otherwise make it so plasmafires operate much faster than wood fires.

The items I added to NewSettings.dm are used to subjectively tweak this sort of thing. I have tried to slow it down even further, but it gets to a point where it feels very inorganic. My goal has been to have fire start slow and build exponentially but it is very difficult to get there using a quasi-realistic system as IRL a fire in a closed off room made of wood like the bar would grow very quickly.


> * I didn't actually check any of the hundreds of changes to flammability values on various items, I'm guessing you did them right.

Yeah and those are easy to fix if oddities are found in game (only have to set flammable = FALSE now instead of worrying about multiple vars).

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		if(prob(clamp(lerp(temp,T20C,T0C + 1000,96,100),96,100))) //4% chance of smoke at 20C, 0% at 1000C
+			return FALSE
+	var/smoke_density = clamp(5 * ((MINOXY2BURN/oxy) ** 2),1,5)
+	var/datum/effect/system/smoke_spread/bad/smoke = new /datum/effect/system/smoke_spread/bad()
```
It doesn't really matter, but you don't have to include the type in the `new()`, since the var is typed

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+/atom/proc/burnSolidFuel()
+	//Don't burn the container until all reagents have been depleted via burnLiquidFuel().
+	if(reagents)
+		if(reagents.total_volume && !istype(src, /obj/item/weapon/reagent_containers/food) && ((locate(/obj/effect/decal/cleanable/liquid_fuel) in src) || has_liquid_fuel()))
```
Man, don't write `istype(src, whatever)`. That's what inheritance is for. Use a proc or var overridden on `/obj/item/weapon/reagent_containers/food` instead of typechecking `src`

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		return
+
+	//Setup
+	var/turf/T = isturf(src) ? src : get_turf(src)
```
You're typechecking `src` again, but you don't even have to change anything else because `get_turf()` works fine on turfs
```suggestion
	var/turf/T = get_turf(src)
```

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+	co2_prod = oxy_used //simplification
+
+	//Start a fire on the tile if a burning object is present without an underlying fire effect.
+	if(!in_fire && T)
```
You don't have to check `T` here, since the proc would have already returned earlier if it were null.

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		//Change in internal energy = change in energy due to heat transfer due to isochoric reaction
+		delta_t = heat_out/(delta_m * material.heating_value)
+		T.hotspot_expose(temperature + delta_t, CELL_VOLUME, surfaces=1)
+		new /obj/effect/fire(loc)
```
I don't think fires like being in anything other than turfs. Also, if `src` is a turf, `loc` is its `area`, which is *definitely* not what you want.

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		extinguish()
+		return
+	var/datum/thermal_material/material = src.thermal_material
+	var/datum/gas_mixture/air = T.return_air()
```
A lot of the code in this file assumes that the burning object isn't in a sealed environment. I don't think that issue is unique to this, though.

west3436 replied:
An exposed environment will still have some air until it's in vacuum - as pressure and mols are factored into the burnrate, I think this is addressed already.

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		return
+
+	if(!reagents.total_volume)
+		return
```
```suggestion
	if(!reagents?.total_volume)
		return
```
If you want.

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+		return
+
+	//Setup
+	var/turf/T = isturf(src) ? src : get_turf(src)
```
Same concerns as above for both of these.

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+	var/in_fire = FALSE
+	for(var/obj/effect/fire/F in T)
+		in_fire = TRUE
+		break
```
You can use `locate()` instead of looping through the turf's contents

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+	var/in_fire = FALSE
+	for(var/obj/effect/fire/F in T)
+		in_fire = TRUE
+		break
```
`locate()` again

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+			reagents.remove_reagent(liquid.id,1) //evaporate non-flammable reagents
+
+	//Start a fire on the tile if a burning object is present without an underlying fire effect.
+	if(!in_fire && T)
```
No need for the check for `T` again

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+	//Start a fire on the tile if a burning object is present without an underlying fire effect.
+	if(!in_fire && T)
+		T.hotspot_expose(max_temperature, CELL_VOLUME, surfaces=1)
+		new /obj/effect/fire(loc)
```
Concern about `loc` again

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
-	if(fire_dmi && fire_sprite)
-		fire_overlay = image(fire_dmi,fire_sprite)
+
+	if(fire_dmi && fire_sprite && !isturf(src) && !fire_overlay)
```
Another `src` typecheck

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
 		overlays += fire_overlay
 
 	var/atom/movable/AM = src
 	if(istype(AM))
```
You didn't write this one, but it's another `src` typecheck. This one's easy to fix, though

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -158,57 +478,102 @@ Attach to transfer valve and open. BOOM.
+
+	for(var/obj/effect/fire/F in src)
+		in_fire = TRUE
+		break
```
Can use `locate()` again

Exxion commented on code/ZAS/Fire.dm:
```diff
@@ -40,115 +132,343 @@ Attach to transfer valve and open. BOOM.
+
+	heat_out *= 1000
+
+	return list("heat_out"=heat_out,"oxy_used"=oxy_used,"co2_prod"=co2_prod,"max_temperature"=material.flame_temp)
```
If this weren't BYOND, I'd recommend using (the equivalent of) a datum for this rather than an associative list, but I'm honestly not sure which is faster in this case. Don't need to change it in this PR, but something to look into

Exxion commented on code/ZAS/_gas_mixture.dm:
```diff
@@ -131,6 +131,8 @@
 	return temperature * heat_capacity()
 
+/datum/gas_mixture/proc/molar_ratio(g) //ratio of moles of the input gas to total mols of all gasses in the area
+	return src[g]/ total_moles
```
```suggestion
	return src[g] / total_moles
```

Exxion commented on code/controllers/subsystem/burnable.dm:
```diff
@@ -42,14 +34,31 @@ var/list/atom/burnableatoms = list()
+		var/in_fire = FALSE
+		for(var/obj/effect/fire/F in loc)
+			in_fire = TRUE
+			break
```
Can use `locate()` again

Exxion commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -31,6 +31,8 @@
 	update_icon()
 	//forceMove(loc) // So that Crossed gets called, so that stacks can be merged
+	initial_thermal_mass = thermal_mass
+	thermal_mass = initial_thermal_mass * amount
```
I think this should be `src.amount` rather than the arg `amount`, which can be null, which would result in this being zero

Exxion commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -268,14 +277,17 @@
 		F.add_fingerprint(user)
 		F.transfer_data_from(src,1)
+		F.thermal_mass = thermal_mass_used
 		use(1)
+		thermal_mass -= thermal_mass_used
```
Shouldn't `use()` itself update `thermal_mass`? It seems very funky that it doesn't

Exxion commented on code/game/objects/items/weapons/cards_ids.dm:
```diff
@@ -16,12 +16,13 @@
-	autoignition_temperature = AUTOIGNITION_PLASTIC
 	quick_equip_priority = list(slot_wear_id)
-
+  
```
Rogue whitespace alert

west3436 replied:
itsloose.png

Exxion commented on code/game/objects/objs.dm:
```diff
@@ -64,13 +64,22 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
+			if(W_CLASS_GIANT)
+				thermal_mass = 100
+	if(thermal_mass)
+		initial_thermal_mass = thermal_mass //can't just use initial() here as some thermal masses are defined in New()
```
Since the supercall is always at the beginning of `New()`, this will be executed before any children update their `thermal_mass` values anyway

west3436 replied:
I'm seeing `initial_thermal_mass` values in testing just fine - do you have a recommended change here?

west3436 replied:
This will have to be kept in mind if adding any objects which define `thermal_mass` in New(), but for now there are no instances where this is an issue.

Exxion commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -60,30 +61,25 @@
+	..()
+	if(amount)
+		var/burnt_papers = round(amount-(thermal_mass - initial_thermal_mass))
+		var/i = 0
+		while(i < burnt_papers)
+			var/obj/item/weapon/paper/P
+			if(papers.len > 0)
+				P = papers[papers.len]
+				papers.Remove(P)
+			amount--
+			i++
```
1. You can use a `for` loop (like `for(var/i in 1 to burnt_papers)`)
2. Removed papers should be `qdel()`ed, right?

Exxion commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -127,6 +123,7 @@
 		to_chat(user, "<span class='notice'>You put [I] in [src].</span>")
 		papers.Add(I)
 		amount++
+		thermal_mass++
```
It doesn't look like this gets reduced when papers are removed

Exxion commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -60,30 +61,25 @@
+		update_icon()
 
 /obj/item/weapon/paper_bin/Exited(atom/movable/Obj, atom/newloc)
 	papers -= Obj
```
You didn't write this, but this doesn't update `amount` *or* `thermal_mass`
It should probably be like
```
if(papers.Remove(Obj))
	amount--
	thermal_mass--
```

Exxion reviewed (changes requested):
You should probably mention the `w_class` changes. I don't think anyone will care, but better to let people know anyway.
`w_type`, too, even though I'm pretty sure I'm the only person who has ever used the recycler

west3436 commented:
> You should probably mention the `w_class` changes. I don't think anyone will care, but better to let people know anyway. `w_type`, too, even though I'm pretty sure I'm the only person who has ever used the recycler

Added to the OP.

jwhitak commented:
Fire fire...
